File: Alcoholtracker/Services/BarcodeService.swift
import Foundation

// MARK: - BarcodeService (B8)
// Looks up a product barcode on Open Food Facts and returns a DrinkTemplateCandidate.

struct DrinkTemplateCandidate {
    enum Source: String {
        case local
        case community
        case openFoodFacts
        case manual

        var label: String {
            switch self {
            case .local: return "Lokal gelernt"
            case .community: return "Community-Datenbank"
            case .openFoodFacts: return "Open Food Facts"
            case .manual: return "Manuelle Eingabe"
            }
        }
    }

    let name: String
    let abv: Double
    let barcode: String
    var volume: Double = 330
    var category: DrinkCategory = .beer
    // false when the barcode was not found in any database and the user is
    // filling the data in by hand; the candidate sheet adapts its wording and
    // the manual entry still feeds the community DB.
    var foundInDatabase: Bool = true
    var source: Source = .openFoodFacts
    var adjustedBySanitizer: Bool = false

    init(
        name: String,
        abv: Double,
        barcode: String,
        volume: Double = 330,
        category: DrinkCategory = .beer,
        foundInDatabase: Bool = true,
        source: Source = .openFoodFacts
    ) {
        self.name = name
        let safeABV = BarcodeService.sanitizedABV(abv)
        let safeVolume = BarcodeService.sanitizedVolumeML(volume)
        let safeCategory = BarcodeService.sanitizedCategory(category, abv: safeABV)
        self.abv = safeABV
        self.barcode = barcode
        self.volume = safeVolume
        self.category = safeCategory
        self.foundInDatabase = foundInDatabase
        self.source = source
        self.adjustedBySanitizer = abs(safeABV - abv) > 0.001
            || abs(safeVolume - volume) > 0.001
            || safeCategory != category
    }
}

enum BarcodeService {

    static func lookup(barcode: String) async throws -> DrinkTemplateCandidate? {
        var lookupCode = barcode
        if lookupCode.count == 12 {
            lookupCode = "0" + lookupCode // Pad UPC-A to EAN-13 for Open Food Facts
        }
        guard let url = URL(string: "https://world.openfoodfacts.org/api/v0/product/\(lookupCode).json") else {
            return nil
        }
        // Open Food Facts asks every client to identify itself with a real
        // User-Agent; generic ones can be throttled or blocked, which would make
        // every lookup fail. https://world.openfoodfacts.org/data
        var request = URLRequest(url: url)
        request.setValue("Promille-App/1.0 (iOS; Getraenke-Tracker)", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = 15

        let (data, response) = try await URLSession.shared.data(for: request)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else { return nil }

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        let statusInt = json?["status"] as? Int
        let statusStr = json?["status"] as? String
        let isSuccess = statusInt == 1 || statusStr == "1"
        // status != 1 means the barcode is genuinely unknown to Open Food Facts.
        guard isSuccess, let product = json?["product"] as? [String: Any] else { return nil }

        let name = (product["product_name"] as? String)?.trimmingCharacters(in: .whitespaces)
            ?? (product["brands"] as? String)?.components(separatedBy: ",").first?.trimmingCharacters(in: .whitespaces)

        guard let productName = name, !productName.isEmpty else { return nil }

        // ABV. Open Food Facts stores the alcohol percentage as "% vol" both in
        // `alcohol_value` and in `nutriments.alcohol_100g` (the _100g suffix is
        // misleading: verified against real data the number IS the % vol, e.g.
        // 1664 -> 5.5, Corona -> 4.5). So use it directly, do NOT divide by any
        // density. Real beers almost always have only alcohol_100g populated.
        func parseNumber(_ any: Any?) -> Double? {
            if let d = any as? Double { return d }
            if let s = any as? String { return Double(s.replacingOccurrences(of: ",", with: ".")) }
            return nil
        }
        let nutriments = product["nutriments"] as? [String: Any]
        let positive: (Double?) -> Double? = { v in (v ?? 0) > 0 ? v : nil }
        // Unknown alcohol -> 0, so the product is still returned and the user can
        // fill in the percentage on the candidate sheet instead of a dead end.
        let abv = positive(parseNumber(product["alcohol_value"]))
            ?? positive(parseNumber(nutriments?["alcohol_100g"]))
            ?? 0

        var quantityParts = [
            product["quantity"] as? String,
            product["product_quantity"] as? String,
            product["serving_size"] as? String
        ].compactMap { $0 }
        if let quantity = parseNumber(product["product_quantity"]) {
            quantityParts.append(String(quantity))
        }
        if let unit = product["product_quantity_unit"] as? String {
            quantityParts.append(unit)
        }
        let quantityText = quantityParts.joined(separator: " ")
        let volume = sanitizedVolumeML(parseVolumeML(from: quantityText) ?? 330)

        var tags: [String] = [productName]
        if let categories = product["categories"] as? String { tags.append(categories) }
        if let genericName = product["generic_name"] as? String { tags.append(genericName) }
        if let categoryTags = product["categories_tags"] as? [String] { tags.append(contentsOf: categoryTags) }
        if let hierarchy = product["categories_hierarchy"] as? [String] { tags.append(contentsOf: hierarchy) }
        let category = sanitizedCategory(inferCategory(from: tags, abv: abv), abv: abv)

        return DrinkTemplateCandidate(
            name: productName,
            abv: abv,
            barcode: barcode,
            volume: volume,
            category: category,
            source: .openFoodFacts
        )
    }

    static func sanitizedABV(_ value: Double) -> Double {
        guard value.isFinite else { return 0 }
        return min(80, max(0, value))
    }

    static func sanitizedVolumeML(_ value: Double) -> Double {
        guard value.isFinite else { return 330 }
        return min(3000, max(5, value))
    }

    static func sanitizedCategory(_ category: DrinkCategory, abv: Double) -> DrinkCategory {
        let safeABV = sanitizedABV(abv)
        if safeABV <= 0.05 {
            return category
        }
        switch category {
        case .water, .softDrink, .juice, .coffeeTea, .milk:
            return .mixed
        default:
            return category
        }
    }

    static func parseVolumeML(from text: String) -> Double? {
        let normalized = text
            .lowercased()
            .replacingOccurrences(of: ",", with: ".")
            .replacingOccurrences(of: "ℓ", with: "l")
        let unitPattern = #"(ml|milliliter|cl|centiliter|l|liter|litre)\b"#
        let packPattern = #"\b\d+\s*[x×]\s*(\d+(?:\.\d+)?)\s*"# + unitPattern
        if let volume = firstVolumeMatch(in: normalized, pattern: packPattern) {
            return volume
        }
        let pattern = #"(\d+(?:\.\d+)?)\s*"# + unitPattern
        return firstVolumeMatch(in: normalized, pattern: pattern)
    }

    private static func firstVolumeMatch(in text: String, pattern: String) -> Double? {
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(in: text, range: NSRange(text.startIndex..., in: text)),
              let valueRange = Range(match.range(at: 1), in: text),
              let unitRange = Range(match.range(at: 2), in: text),
              let value = Double(text[valueRange])
        else { return nil }

        let unit = String(text[unitRange])
        if unit.hasPrefix("ml") || unit.hasPrefix("milli") { return value }
        if unit.hasPrefix("cl") || unit.hasPrefix("centi") { return value * 10 }
        return value * 1000
    }

    private static func inferCategory(from tags: [String], abv: Double) -> DrinkCategory {
        let text = tags.joined(separator: " ").lowercased()
        func has(_ needles: String...) -> Bool { needles.contains { text.contains($0) } }

        if has("wine", "wein", "vino", "vin ") { return .wine }
        if has("sparkling", "champagne", "sekt", "prosecco", "cava") { return .sparkling }
        if has("spirit", "whisky", "whiskey", "vodka", "wodka", "rum", "gin", "tequila", "schnaps") { return .spirits }
        if has("liqueur", "likör", "likoer") { return .liqueur }
        if has("cider") { return .cider }
        if has("beer", "bier", "cerveza", "biere") { return .beer }
        if has("water", "wasser", "mineral-water", "eau-minerale", "still-water", "sparkling-water") { return .water }
        if has("juice", "saft", "jus", "nectar", "smoothie") { return .juice }
        if has("coffee", "kaffee", "café", "tea", "tee", "iced-tea", "eistee") { return .coffeeTea }
        if has("milk", "milch", "yoghurt-drink", "kefir") { return .milk }
        if has("soda", "soft-drink", "soft drink", "cola", "lemonade", "limonade", "energy-drink", "isotonic") { return .softDrink }
        return abv > 0 ? .beer : .softDrink
    }
}

File: android/app/src/main/kotlin/de/tipau/promille/service/BarcodeService.kt
package de.tipau.promille.service

import de.tipau.promille.bac.DrinkCategory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.regex.Pattern
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

enum class CandidateSource(val label: String) {
    LOCAL("Lokal gelernt"),
    COMMUNITY("Community-Datenbank"),
    OPEN_FOOD_FACTS("Open Food Facts"),
    MANUAL("Manuelle Eingabe")
}

data class DrinkTemplateCandidate(
    val name: String,
    val abv: Double,
    val barcode: String,
    val volume: Double = 330.0,
    val category: DrinkCategory = DrinkCategory.BEER,
    val foundInDatabase: Boolean = true,
    val source: CandidateSource = CandidateSource.OPEN_FOOD_FACTS,
    val adjustedBySanitizer: Boolean = false
) {
    companion object {
        fun create(
            name: String,
            abv: Double,
            barcode: String,
            volume: Double = 330.0,
            category: DrinkCategory = DrinkCategory.BEER,
            foundInDatabase: Boolean = true,
            source: CandidateSource = CandidateSource.OPEN_FOOD_FACTS
        ): DrinkTemplateCandidate {
            val safeABV = BarcodeService.sanitizedABV(abv)
            val safeVolume = BarcodeService.sanitizedVolumeML(volume)
            val safeCategory = BarcodeService.sanitizedCategory(category, safeABV)
            val adjusted = abs(safeABV - abv) > 0.001 ||
                abs(safeVolume - volume) > 0.001 ||
                safeCategory != category

            return DrinkTemplateCandidate(
                name = name,
                abv = safeABV,
                barcode = barcode,
                volume = safeVolume,
                category = safeCategory,
                foundInDatabase = foundInDatabase,
                source = source,
                adjustedBySanitizer = adjusted
            )
        }
    }
}

object BarcodeService {

    suspend fun lookup(barcode: String): DrinkTemplateCandidate? = withContext(Dispatchers.IO) {
        var lookupCode = barcode.trim()
        if (lookupCode.length == 12) {
            lookupCode = "0$lookupCode" // Pad UPC-A to EAN-13 for Open Food Facts
        }

        try {
            val url = URL("https://world.openfoodfacts.org/api/v0/product/$lookupCode.json")
            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.setRequestProperty("User-Agent", "Promille-App/1.0 (Android; Getraenke-Tracker)")
            connection.connectTimeout = 15000
            connection.readTimeout = 15000

            if (connection.responseCode != 200) {
                return@withContext null
            }

            val responseBody = connection.inputStream.bufferedReader().use { it.readText() }
            val json = JSONObject(responseBody)
            val status = json.optInt("status", -1)
            if (status != 1 || !json.has("product")) {
                return@withContext null
            }

            val product = json.getJSONObject("product")
            val productName = product.optString("product_name").trim().ifEmpty {
                val brands = product.optString("brands").trim()
                if (brands.isNotEmpty()) brands.split(",").first().trim() else ""
            }

            if (productName.isEmpty()) return@withContext null

            // ABV extraction
            var abv = 0.0
            if (product.has("alcohol_value")) {
                abv = product.optDouble("alcohol_value", 0.0)
            }
            if (abv <= 0.0 && product.has("nutriments")) {
                val nutriments = product.getJSONObject("nutriments")
                abv = nutriments.optDouble("alcohol_100g", 0.0)
                if (abv <= 0.0) {
                    abv = nutriments.optDouble("alcohol", 0.0)
                }
            }
            if (abv.isNaN() || abv < 0.0) abv = 0.0

            // Quantity / volume extraction
            val quantityParts = mutableListOf<String>()
            product.optString("quantity").takeIf { it.isNotBlank() }?.let { quantityParts.add(it) }
            product.optString("product_quantity").takeIf { it.isNotBlank() }?.let { quantityParts.add(it) }
            product.optString("serving_size").takeIf { it.isNotBlank() }?.let { quantityParts.add(it) }
            product.optString("product_quantity_unit").takeIf { it.isNotBlank() }?.let { quantityParts.add(it) }

            val quantityText = quantityParts.joinToString(" ")
            val volume = sanitizedVolumeML(parseVolumeML(quantityText) ?: 330.0)

            // Category inference
            val tags = mutableListOf<String>()
            tags.add(productName)
            product.optString("categories").takeIf { it.isNotBlank() }?.let { tags.add(it) }
            product.optString("generic_name").takeIf { it.isNotBlank() }?.let { tags.add(it) }

            val categoryTags = product.optJSONArray("categories_tags")
            if (categoryTags != null) {
                for (i in 0 until categoryTags.length()) {
                    tags.add(categoryTags.optString(i))
                }
            }

            val rawCategory = inferCategory(tags, abv)
            val category = sanitizedCategory(rawCategory, abv)

            DrinkTemplateCandidate.create(
                name = productName,
                abv = abv,
                barcode = barcode,
                volume = volume,
                category = category,
                source = CandidateSource.OPEN_FOOD_FACTS
            )
        } catch (_: Exception) {
            null
        }
    }

    fun sanitizedABV(value: Double): Double {
        if (!value.isFinite()) return 0.0
        return min(80.0, max(0.0, value))
    }

    fun sanitizedVolumeML(value: Double): Double {
        if (!value.isFinite()) return 330.0
        return min(3000.0, max(5.0, value))
    }

    fun sanitizedCategory(category: DrinkCategory, abv: Double): DrinkCategory {
        val safeABV = sanitizedABV(abv)
        if (safeABV <= 0.05) return category
        return when (category) {
            DrinkCategory.WATER,
            DrinkCategory.SOFT_DRINK,
            DrinkCategory.JUICE,
            DrinkCategory.COFFEE_TEA,
            DrinkCategory.MILK -> DrinkCategory.MIXED
            else -> category
        }
    }

    fun parseVolumeML(text: String): Double? {
        val normalized = text.lowercase()
            .replace(",", ".")
            .replace("ℓ", "l")

        val unitPattern = "(ml|milliliter|cl|centiliter|l|liter|litre)\\b"
        val packPattern = "\\b\\d+\\s*[x×]\\s*(\\d+(?:\\.\\d+)?)\\s*$unitPattern"
        firstVolumeMatch(normalized, packPattern)?.let { return it }

        val pattern = "(\\d+(?:\\.\\d+)?)\\s*$unitPattern"
        return firstVolumeMatch(normalized, pattern)
    }

    private fun firstVolumeMatch(text: String, patternString: String): Double? {
        val matcher = Pattern.compile(patternString).matcher(text)
        if (!matcher.find()) return null
        val valueStr = matcher.group(1) ?: return null
        val unit = matcher.group(2) ?: return null
        val value = valueStr.toDoubleOrNull() ?: return null

        return when {
            unit.startsWith("ml") || unit.startsWith("milli") -> value
            unit.startsWith("cl") || unit.startsWith("centi") -> value * 10.0
            else -> value * 1000.0
        }
    }

    private fun inferCategory(tags: List<String>, abv: Double): DrinkCategory {
        val text = tags.joinToString(" ").lowercase()
        fun has(vararg needles: String): Boolean = needles.any { text.contains(it) }

        if (has("wine", "wein", "vino", "vin ")) return DrinkCategory.WINE
        if (has("sparkling", "champagne", "sekt", "prosecco", "cava")) return DrinkCategory.SPARKLING
        if (has("spirit", "whisky", "whiskey", "vodka", "wodka", "rum", "gin", "tequila", "schnaps")) return DrinkCategory.SPIRITS
        if (has("liqueur", "likör", "likoer")) return DrinkCategory.LIQUEUR
        if (has("cider")) return DrinkCategory.CIDER
        if (has("beer", "bier", "cerveza", "biere")) return DrinkCategory.BEER
        if (has("water", "wasser", "mineral-water", "eau-minerale", "still-water", "sparkling-water")) return DrinkCategory.WATER
        if (has("juice", "saft", "jus", "nectar", "smoothie")) return DrinkCategory.JUICE
        if (has("coffee", "kaffee", "café", "tea", "tee", "iced-tea", "eistee")) return DrinkCategory.COFFEE_TEA
        if (has("milk", "milch", "yoghurt-drink", "kefir")) return DrinkCategory.MILK
        if (has("soda", "soft-drink", "soft drink", "cola", "lemonade", "limonade", "energy-drink", "isotonic")) return DrinkCategory.SOFT_DRINK
        return if (abv > 0) DrinkCategory.BEER else DrinkCategory.SOFT_DRINK
    }
}
